### PR TITLE
feat(quota): add Request and Refresh middlewares for APIs

### PR DESCRIPTION
**Pull request by heww**
1. Introduce ReqquestMiddleware and RefereshMiddleware.
2. Add request middlware to copy artifact, mount blob, put blob upload,
put manifest, upload chart verson APIs.
3. Add refresh project middleware to delete manifest, delete artifact,
delete chart version, delete repository APIs.

Signed-off-by: He Weiwei <hweiwei@vmware.com>

### CODE REVIEW

File: src/api/blob/controller.go [hunk 1]
```diff
@@ -188,6 +160,69 @@ func (c *controller) Exist(ctx context.Context, digest string, options ...Option
...
+	for _, blob := range blobs {
+		if !associated[blob.ID] {
+			results = append(results, blob)
+		}
+	}
+
+	return results, nil
+}
+
+func (c *controller) Get(ctx context.Context, digest string, options ...Option) (*blob.Blob, error) {
```
Comment (wy65701436) [inline src/api/blob/controller.go]: if I only pass the digest, what blob should I get?

File: src/server/middleware/blob/copy_artifact.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	"github.com/goharbor/harbor/src/server/middleware"
+	"github.com/goharbor/harbor/src/server/middleware/util"
+)
+
+func isSuccess(statusCode int) bool {
+	return statusCode >= http.StatusOK && statusCode < http.StatusBadRequest
+}
+
+// CopyArtifactMiddleware middleware to sync the missing associations for the project
+func CopyArtifactMiddleware() func(http.Handler) http.Handler {
```
Comment (wy65701436) [inline src/server/middleware/blob/copy_artifact.go]: does the error code meet the OCI spec?

File: src/server/v2.0/handler/handler.go [hunk 3]
```diff
@@ -31,6 +35,10 @@ func New() http.Handler {
 		log.Fatal(err)
 	}
 
+	api.RegisterMiddleware("CopyArtifact", middleware.Chain(quota.CopyArtifactMiddleware(), blob.CopyArtifactMiddleware()))
```
Comment (wy65701436) [inline src/server/v2.0/handler/handler.go]: is it possible to merge these two middlewares? They share some queries, I worried about the performance impact.

Comment (heww) [reply src/server/v2.0/handler/handler.go]: This are two things, quota can be disabled, blob can not be disabled , gc depent on blobs.

Performance problem can be solved by other methods.

File: src/server/middleware/blob/copy_artifact.go (hunk 2 above)
Comment (heww) [reply src/server/middleware/blob/copy_artifact.go]: I think so, the `err` is processed by `SendError` in the `server/error` pkg.

File: src/api/blob/controller.go (hunk 1 above)
Comment (heww) [reply src/api/blob/controller.go]: The blob of the digest itself.

File: src/server/middleware/quota/copy_artifact.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+		log.Errorf("%s: parse from parameter failed, error: %v", logPrefix, err)
+		return nil, nil
+	}
+
+	ctx := r.Context()
+
+	art, err := artifactController.GetByReference(ctx, repository, reference, nil)
+	if ierror.IsNotFoundErr(err) {
+		// artifact not found, discontinue the API request
+		return nil, ierror.BadRequestError(nil).WithMessage("artifact %s not found", from)
```
Comment (ywk253100) [inline src/server/middleware/quota/copy_artifact.go]: Seems you can return error directly without creating a new one

File: src/pkg/blob/dao/dao.go [hunk 5]
```diff
@@ -167,12 +172,34 @@ func (d *dao) UpdateBlob(ctx context.Context, blob *models.Blob) error {
...
-func (d *dao) ListBlobs(ctx context.Context, query *q.Query) ([]*models.Blob, error) {
-	qs, err := orm.QuerySetter(ctx, &models.Blob{}, query)
+func (d *dao) ListBlobs(ctx context.Context, params models.ListParams) ([]*models.Blob, error) {
+	qs, err := orm.QuerySetter(ctx, &models.Blob{}, nil)
 	if err != nil {
 		return nil, err
 	}
 
+	if len(params.BlobDigests) > 0 {
+		qs = qs.Filter("digest__in", params.BlobDigests)
```
Comment (ywk253100) [inline src/pkg/blob/dao/dao.go]: Replace all `__in` and other Beego ORM operator if any with the new query model

File: src/server/middleware/quota/copy_artifact.go (hunk 4 above)
Comment (heww) [reply src/server/middleware/quota/copy_artifact.go]: This change the response from 404 to 400.

File: src/pkg/blob/dao/dao.go (hunk 5 above)
Comment (heww) [reply src/pkg/blob/dao/dao.go]: I make `QuerySetter` support `xxx__in` key in `q.KeyWords`.

**Pull request merged**